### ISSUE TITLE
Handle cross-thread server shutdown exceptions

### ISSUE THREAD

**Issue by jaredbroad** [rabbitmq/rabbitmq-dotnet-client#336]
We use RMQ as a critical part of our infrastructure and have an issue that if the server is shutdown for any reason all bound consumers have fatal exceptions killing the application. For our use case this is devastating as we need 100% uptime and the programs have delicate state. 

Recently a server update shutdown the RMQ host and caused hundreds of clients to have this exception =(. Its a pretty horrible situation and would really appreciate any advice on how to handle it:

```
2017-07-07T02:25:23.5885250Z ERROR:: Unhandled Exception:
2017-07-07T02:25:23.5891520Z ERROR:: System.ObjectDisposedException: Cannot access a disposed object.
2017-07-07T02:25:23.5895510Z ERROR:: Object name: 'handle'.
2017-07-07T02:25:23.5898440Z ERROR:: at System.Runtime.InteropServices.SafeHandle.DangerousAddRef (System.Boolean& success) [0x00025] in <dca3b561b8ad4f9fb10141d81b39ff45>:0
2017-07-07T02:25:23.5899830Z ERROR:: at System.Threading.NativeEventCalls.SetEvent (Microsoft.Win32.SafeHandles.SafeWaitHandle handle) [0x00002] in <dca3b561b8ad4f9fb10141d81b39ff45>:0
2017-07-07T02:25:23.5901030Z ERROR:: at System.Threading.EventWaitHandle.Set () [0x00018] in <dca3b561b8ad4f9fb10141d81b39ff45>:0
2017-07-07T02:25:23.5902040Z ERROR:: at (wrapper remoting-invoke-with-check) System.Threading.EventWaitHandle:Set ()
2017-07-07T02:25:23.5903030Z ERROR:: at System.IOAsyncResult.set_IsCompleted (System.Boolean value) [0x00024] in <59be416de143456b88b9988284f43350>:0
2017-07-07T02:25:23.5903940Z ERROR:: at System.Net.Sockets.SocketAsyncResult.Complete () [0x00044] in <59be416de143456b88b9988284f43350>:0
2017-07-07T02:25:23.5905020Z ERROR:: at System.Net.Sockets.SocketAsyncResult.Complete (System.Exception e) [0x00007] in <59be416de143456b88b9988284f43350>:0
2017-07-07T02:25:23.5907040Z ERROR:: at System.Net.Sockets.Socket.<BeginConnectCallback>m__4 (System.IOAsyncResult ares) [0x0012a] in <59be416de143456b88b9988284f43350>:0
2017-07-07T02:25:23.5907980Z ERROR:: at System.IOSelectorJob.System.Threading.IThreadPoolWorkItem.ExecuteWorkItem () [0x00000] in <59be416de143456b88b9988284f43350>:0
2017-07-07T02:25:23.5908730Z ERROR:: at System.Threading.ThreadPoolWorkQueue.Dispatch () [0x00096] in <dca3b561b8ad4f9fb10141d81b39ff45>:0
2017-07-07T02:25:23.5909540Z ERROR:: at System.Threading._ThreadPoolWaitCallback.PerformWaitCallback () [0x00000] in <dca3b561b8ad4f9fb10141d81b39ff45>:0
```

It was using Mono 4.6.15 with RMQ Client version  3.5.3 -- we will update to the latest version of RMQ now -- is there anything else we can do? Is there an event handler for these cases?

**Comment by michaelklishin** [comment]
Thank you for your time.

Team RabbitMQ uses GitHub issues for specific actionable items engineers can work on. This assumes two things:

1. GitHub issues are not used for questions, investigations, root cause analysis, discussions of potential issues, etc (as defined by this team)
2. We have a certain amount of information to work with

We get at least a dozen of questions through various venues every single day, often quite light on details.
At that rate GitHub issues can very quickly turn into a something impossible to navigate and make sense of even for our team. Because of that questions, investigations, root cause analysis, discussions of potential features are all considered to be mailing list material by our team. Please post this to [rabbitmq-users](https://groups.google.com/forum/#!forum/rabbitmq-users).

Getting all the details necessary to reproduce an issue, make a conclusion or even form a hypothesis about what's happening can take a fair amount of time. Our team is multiple orders of magnitude smaller than the RabbitMQ community. Please help others help you by providing a way to reproduce the behavior you're
observing, or at least sharing as much relevant information as possible [on the list](https://groups.google.com/forum/#!forum/rabbitmq-users):

 * Server, client library and plugin (if applicable) versions used
 * Server logs
 * A code example or terminal transcript that can be used to reproduce
 * **Full** exception stack traces (not a single line message)
 * `rabbitmqctl status` (and, if possible, `rabbitmqctl environment` output)
 * Other relevant things about the environment and workload, e.g. [a traffic capture](https://www.rabbitmq.com/amqp-wireshark.html)

Feel free to edit out hostnames and other potentially sensitive information.

When/if we have enough details and evidence we'd be happy to file a new issue.

Thank you.

**Comment by michaelklishin** [comment]
@jaredbroad I'm not sure if this is the entire stack trace but much of it ends up being system libraries.
There is no evidence that it is RabbitMQ .NET client that cause Mono runtime to terminate and even if it was, shouldn't the runtime be reasonably defensive and not exit when a bunch of socket exceptions aren't handled?

Ever since we introduced .NET Core support, we no longer test this client on Mono (at least not directly). This client is approaching a 5.0 release. You should be able to use 4.x or even 5.0 against RabbitMQ 3.5.3.

**Comment by michaelklishin** [comment]
#220, #223, #230 and #263 are potentially relevant. #220 was reported against `3.5.4`, in fact.

**Comment by kjnilsson** [comment]
I'd recommend trying the latest 4.x version. Those code-paths have changed quote a bit.

**Comment by jaredbroad** [comment]
Thank you @michaelklishin @kjnilsson 

We've started the upgrade process; its a very reproducible issue as we went to great lengths to ensure the servers were never rebooted (7+months uptime..). 

From what I understand of #220 and the patches for it that saves the unhandled exception on shutdown -- we set the connections to auto recover so it was surprising to still see the issue - the automatic recovery worked well in Windows testing but not in Mono. 

Should we expect it to shutdown (cleanly in 4.x) when auto-recover is true? Or will it continue attempting to connect (as we hoped).
```
var factory = new ConnectionFactory
            {
                ...
                AutomaticRecoveryEnabled = true,
                RequestedConnectionTimeout = 5000
            };
```
